[PATCH v10] api: crypto: copy the ranges of null algorithms in OOP op type & change bit mode semantics

JannePeltonen opened the pull request:
Allow non-zero-length cipher and auth ranges for the null cipher and auth algoriths in the OOP operation type. Copy the ranges to the output packet as with non-null algorithms. Allow implementations to not fully support non-zero-length ranges by explicitly mentioning that a crypto operation with such a range may fail if not supported.

This changes makes null algorithms behave as non-null algorithms with respect to packet data copying in the OOP operation type. When both the cipher and the auth algorithms are null, a crypto operation reduces to a´packet data copying operation (if supported by the underlying ODP implementation). Such copying may be useful for applications that wish to process crypto and non-crypto packets in the same way using the OOP crypto operation type.

JannePeltonen commented:
v2: Added a new API patch that makes bit/byte mode of an algorithm configurable at session creation and changes bit_mode capability to tell if bit mode is supported (as opposed to telling if the algorithm must be used in bit mode in the particular ODP implementation).

This change simplifies the API and avoids the mess that would result from several current ODP implementations advertising both bit mode and byte mode for null algorithms (so an implementation does not know if a range given is in bits or bytes) and then the previously sent out-of-place processing change actually needing well-defined ranges for null algorithms too.

This new API patch, as now proposed, is not backward compatible for applications that may use bit mode algorithms.

anoobj commented:
Spec change looks good.

Reviewed-by: Anoob Joseph <anoobj@marvell.com>

psavol commented on include/odp/api/spec/crypto_types.h:
```diff
@@ -937,22 +967,23 @@ typedef struct odp_crypto_cipher_capability_t {
-	 * which is not divisible by 8 is unsupported and will result in
-	 * unspecified behaviour.
+	 * Note: In bit mode the cipher range must start on a byte boundary.
+	 * using an offset which is not divisible by 8 will result in
```
using -> Using (upper case)

JannePeltonen replied:
Fixed in v4.

psavol commented on include/odp/api/spec/crypto_types.h:
```diff
@@ -984,22 +1015,23 @@ typedef struct odp_crypto_auth_capability_t {
-	 * which is not divisible by 8 is unsupported and will result in
-	 * unpredictable behaviour.
+	 * Note: In bit mode the auth range must start on a byte boundary.
+	 * using an offset which is not divisible by 8 will result in
```
using -> Using (upper case)

JannePeltonen replied:
Fixed in v4.

psavol commented on include/odp/api/spec/crypto_types.h:
```diff
@@ -753,9 +753,11 @@ typedef struct odp_crypto_packet_op_param_t {
+	 *  range is copied in the output packet even with the null cipher.
+	 *  If an ODP implementation does not support non-zero-length
+	 *  ranges with the null cipher and the OOP operation type, the
+	 *  crypto operation may fail.
```
Would it be more precise to say "... operation may report failure.". I.e. to be explicit that operation return code reports a failure rather than there would be undefined behavior.

JannePeltonen replied:
I did not change this yet since we probably need to clarify it more than that. Specifically, if we allow failure, do we allow a failure of odp_crypto_op() / odp_crypto_op_enq() so that the input packet is not even consumed, or a failure reported through the crypto op result, or both. Let's discuss it today.

JannePeltonen replied:
Reworded in v5.

JannePeltonen commented:
v4:
	- Added couple of bug fixes and minor code improvements
	- Added linux-gen implementantion
	- Added validation tests (without allowing a failure of OOP operations with null algorithms and non-zero ranges)

JannePeltonen commented:
v5:
- Reworded the text about allowing OOP crypto op failures with null algorithms and non-zero ranges.
- Added a patch to allow the above failures in validation test without causing test failures

psavol commented:
Reviewed and approved API patches in v5.

JereLeppanen commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -254,15 +254,16 @@ static void alg_test_ses(odp_crypto_op_t op,
 {
```
Typos in commit comment: "resuls", "aurh", "tange".

JereLeppanen commented on test/validation/api/crypto/crypto_op_test.c:
```diff
@@ -179,8 +179,8 @@ static void prepare_crypto_ranges(const crypto_op_test_param_t *param,
 				  odp_packet_data_range_t *auth_range)
```
Typo in commit comment: "parametets".

JannePeltonen commented:
Separated the three first commits to PR 1862 so they are no longer in the scope of this PR except as dependencies until PR 1862 gets merged and this PR rebased.

JannePeltonen commented:
v6: Rebased to get rid of the dependent commits. Now all the commits are part of this PR review.
v7: Added Petri's reviewed-by tags and fixed a typo in a commit message

anoobj commented on platform/linux-generic/odp_crypto_openssl.c:
```diff
@@ -1119,70 +1120,72 @@ static int process_cipher_param(odp_crypto_generic_session_t *session,
+						const odp_crypto_packet_op_param_t *param,
+						odp_crypto_generic_session_t *session)
+{
+	odp_crypto_packet_op_param_t new_param = *param;
```
Can we avoid this copy by passing offset & length as separate args to 'cipher_encrypt_bytes'?

JannePeltonen replied:
Since the prototype of the generic cipher function cannot be easily changed, that would require moving the body of cipher_encrypt_bytes() into a separate function and leaving cipher_encrypt() bytes to be a simpler wrapper. And the same for the decrypt function.

I rather wanted to keep the change simpler. Copying the parameter struct should not be that costly anyway. A better optimization would be to get rid of the packet data copy that occurs in cipher_encrypt_bytes by using internal_encrypt() that can handle also segmented packets without copying. But I did not want to do such a change yet in this PR which is about the bit mode semantics change.

Also, supporting OOP without the separate memory copying step would probably be a good optimization but that would require changing the prototype of the generic cipher & auth functions (and would also otherwise be a bit bigger change). Because of this, I did not want to start changing the generic cipher & auth function prototypes now.

anoobj replied:
Understood. Thanks for the explanation.

JannePeltonen commented on include/odp/api/spec/crypto_types.h:
```diff
@@ -560,6 +560,30 @@ typedef struct odp_crypto_session_param_t {
 	 */
 	odp_crypto_op_type_t op_type;
 
+	/** Units of cipher range
```
Changed to "Cipher range unit" in v8.

JannePeltonen commented on include/odp/api/spec/crypto_types.h:
```diff
@@ -560,6 +560,30 @@ typedef struct odp_crypto_session_param_t {
+	 *  Otherwise the offset and length are in bytes.
+	 *
+	 *  If cipher capabilities do not include bit_mode, setting this flag
+	 *  causes session creation failure.
```
Reworded slightly in v8.

JannePeltonen commented on include/odp/api/spec/crypto.h:
```diff
@@ -255,9 +255,7 @@ int odp_crypto_result(odp_crypto_packet_result_t *result,
  * input packet to the output packet and then the crypto operation
```
fixed in v8

JannePeltonen commented on include/odp/api/spec/crypto_types.h:
```diff
@@ -753,9 +753,15 @@ typedef struct odp_crypto_packet_op_param_t {
-	 *  operation type.
+	 *  ODP_CRYPTO_OP_TYPE_OOP.
+	 *
+	 *  With the OOP operation type the cipher range is copied in the
```
fixed in v8

JannePeltonen commented:
v8: fixed a few typos and grammar issues.

anoobj commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -171,8 +171,10 @@ static odp_crypto_session_t session_create(odp_crypto_op_t op,
+	ses_params.op_type = session->op_type;
+	ses_params.cipher_range_in_bits = session->cipher_range_in_bits;
+	ses_params.auth_range_in_bits = session->auth_range_in_bits;
 	ses_params.auth_cipher_text = (order == AUTH_CIPHERTEXT);
```
Does it make sense to move order also to crypto_session_t?

JannePeltonen replied:
Perhaps, and the same for hash_mode, but since the change would not really be related to any of the commits in this PR it does not need to be in this PR.

anoobj replied:
Agreed.

anoobj commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -342,8 +340,47 @@ static void alg_test(odp_crypto_op_t op,
+	odp_packet_data_range_t cipher_range;
+	odp_packet_data_range_t auth_range;
+
+	for (int cr_in_bits = 0; cr_in_bits <= 1; cr_in_bits++) {
```
Is it right to use 'for' loop here? Isn't it better to introduce a wrapper function like,

alg_test_flags(..., bool is_cr_in_bit, bool is_ar_in_bit)

And then have calls from here like,

alg_test_flags(...., false, false);
alg_test_flags(...., false, true);
alg_test_flags(...., true, false);
alg_test_flags(...., true, true);

Other option is to use goto,

	bool ar_in_bytes, cr_in_bytes = true;

reset_auth_test_again:
	ar_in_bytes = true;

test_again:
	cipher_range = cipher_bit_range;
	auth_range = auth_bit_range;

	if (cr_in_bytes) {
		cipher_range.offset /= 8;
		cipher_range.length /= 8;
	}

	if (ar_in_bytes) {
		auth_range.offset /= 8;
		auth_range.length /= 8;
	}

	alg_test_op_types(op, order, ref, cipher_range, auth_range,
			  digest_offset, cr_in_bits, ar_in_bits);

	if (is_bit_mode_auth && ar_in_bytes) {
		ar_in_bytes = false;
		goto test_again;
	}

	if (is_bit_mode_cipher && cr_in_bytes) {
		cr_in_bytes = false;
		goto reset_auth_test_again;
	}

If we are proceeding with 'for' loop, please check if it is all right to declare variable in for loop header.

JannePeltonen replied:
I do not think there is anything wrong using for statements here. We are iterating over the possible values of the two-valued conditions. I think the example using gotos for the looping is much worse. The wrapper function option would probably be ok, but it essentially just unrolls the for loops, which maybe is not so useful.

Declaring loop variables within the for statement like that has been supported since C99.

I am not planning to change this piece of code unless you insist.

anoobj replied:
My only concern is the code legibility. You can proceed with the current code if you think that is the best way.

anoobj commented:
Minor comment in one of the test patch. Other than that,
Reviewed-by: Anoob Joseph <anoobj@marvell.com>

JannePeltonen commented:
v9: added reviewed-by tags
